cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java

```diff
@@ -250,4 +263,116 @@ private static String getFailMessage(@NotNull Context ctx,
         return "Request for: " + ctx.req.getRequestURI() + " denied. Missing roles: " + missing;
     }
 
+
+    /**
+     * Return create an ApiKey for future authentication and authorization.
+     * @param p Principal object, to get the username
+     * @param keyName Friendly name for this key
+     * @param expires when this key expires; can be null to never expire
+     * @return The created ApiKey
+     */
+    public ApiKey createApiKey(DataApiPrincipal p, ApiKey sourceData) throws CwmsAuthException {
+        
+        try {
+            if(!p.getName().equalsIgnoreCase(sourceData.getUserId())) {
+                throw new CwmsAuthException(ONLY_OWN_KEY_MESSAGE, HttpCode.UNAUTHORIZED.getStatus());
+            }
+            SecureRandom randomSource = SecureRandom.getInstanceStrong();
+            String key = randomSource.ints(48,122) // allow a-zA-Z0-9
+                                 .filter(i -> (i <= 57 || i >= 65) && (i <= 90 || i >= 97))
+                                 .limit(256)
+                                 .collect(StringBuilder::new,StringBuilder::appendCodePoint, StringBuilder::append)
+                                 .toString();
+            final ApiKey newKey = new ApiKey(sourceData.getUserId().toUpperCase(),sourceData.getKeyName(),key,ZonedDateTime.now(ZoneId.of("UTC")),sourceData.getExpires());
+            dsl.connection(c->{
+                setSessionForAuthCheck(c);
+                try (PreparedStatement createKey = c.prepareStatement(CREATE_API_KEY);) {
+                    createKey.setString(1,newKey.getUserId());
+                    createKey.setString(2,newKey.getKeyName());
+                    createKey.setString(3,newKey.getApiKey());
+                    createKey.setDate(4,new Date(newKey.getCreated().toInstant().toEpochMilli()));
```

Comment:
The date will be interpreted in the JVM's timezone. Pass a Calendar object for the 3rd argument to enforce UTC interpretation.
For all set/getDate calls.
https://docs.oracle.com/javase/8/docs/api/java/sql/PreparedStatement.html#setDate-int-java.sql.Date-java.util.Calendar-

Reply:
Thanks. that likely would've bit me a some point.